feat: add products per page dropdown listing

MelvinAchterhuis opened the pull request:
### 1. Why is this change necessary?

Adds more flexibility in displaying products for visitors, easier to compare products / browse through larger catalogues. 

### 2. What does this change do, exactly?

Adds a dropdown next to the sorting where the visitor can select how many products needs to be displayed in the listing

https://github.com/user-attachments/assets/5e94b39f-4daf-43d1-9084-221b1ba0c9a9

To discuss:
- How to make configurable / manageable
- How to prevent visitors from manipulating limit parameter in request to prevent obscure requests

### 3. Describe each step to reproduce the issue or behaviour.
N/A

### 4. Please link to the relevant issues (if any).
N/A

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfill them.

MelvinAchterhuis committed:
feat: add products per page dropdown listing

aragon999 commented:
Very nice :-) maybe a solution of how to make it configurable would be to change the `productsPerPage` to something like `6|12|24` (or add a new one), and then `explode` this and filter it and convert it to `int`. If there is only one value, this is the default, otherwise show the selection.

Then the question is, how to configure the default limit (or should it always be the smallest one).

MelvinAchterhuis commented:
> Very nice :-) maybe a solution of how to make it configurable would be to change the `productsPerPage` to something like `6|12|24` (or add a new one), and then `explode` this and filter it and convert it to `int`. If there is only one value, this is the default, otherwise show the selection.
> 
> Then the question is, how to configure the default limit (or should it always be the smallest one).

I think the default should be what is configured here:
![image](https://github.com/user-attachments/assets/872c501e-8079-4722-ab9e-ecbd1ff82cad)

As for the steps in the dropdown, for a first iteration I was thinking on something like this

- Hidden system config in database with simple steps like 12, 24, 36, 48, 60, 72, 84, 96
- Show active / inactive switch in settings

BrocksiNet commented:
We like the idea 👍🏼 

But maybe we also need the option to overwrite it here (?):
![image](https://github.com/user-attachments/assets/4e2eda9c-c631-4832-a1b2-2a5c44fae893)

I suggest adding a new config field with an empty default value. If someone wants to use the new feature, they can add their own config, like @aragon999, like `6|12|24`. So the default value field stays the same. The question is, what happens when you have the default `32` and this value is missing in the new field?

MelvinAchterhuis commented:
> We like the idea 👍🏼
> 
> But maybe we also need the option to overwrite it here (?): ![image](https://private-user-images.githubusercontent.com/3763023/441114430-4e2eda9c-c631-4832-a1b2-2a5c44fae893.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2MDgyNTgsIm5iZiI6MTc0NjYwNzk1OCwicGF0aCI6Ii8zNzYzMDIzLzQ0MTExNDQzMC00ZTJlZGE5Yy1jNjMxLTQ4MzItYTFiMi0yYTVjNDRmYWU4OTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MDdUMDg1MjM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTNhMjAyODIzMDcxY2YyNzI2NTliMWUxYjVlYzk4NWZmNWJlZjJmZjVmNmE5ZTU0MzY1MGY0ODQ5Y2FjNmFlNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ._2PqIC_IPQqyPcM-k8sHU9dySmAJW1cPDQJZ6VsU1zs)
> 
> I suggest adding a new config field with an empty default value. If someone wants to use the new feature, they can add their own config, like @aragon999, like `6|12|24`. So the default value field stays the same. The question is, what happens when you have the default `32` and this value is missing in the new field?

Thanks for the feedback, I will have a look at this in the next couple of days :)

MelvinAchterhuis committed:
Merge branch 'trunk' into feat/add-products-per-page-dropdown-listing

MelvinAchterhuis committed:
feat: implement element configuration, move logic to resolver

MelvinAchterhuis commented:
Changed the following
- Added configuration for `product-listing` element
- Added system config under `Settings -> products`
- Implemented logic in corresponding resolver
  - Use element configuration or system config
  - Explode by `|`
  - Merge default products per page system config into array
  - Unique + sort
- Add array to slot config to use in Twig

TODO: How to handle limit handling in `PagingListingProcessor.php`, see comment in file

mitelg commented on src/Core/Content/Product/Cms/ProductListingCmsElementResolver.php:
```diff
@@ -168,4 +172,23 @@ private function restrictFilters(CmsSlotEntity $slot, Request $request): void
+        }
+
+
+        $limits = explode('|', $elementLimits ?? $systemLimits ?? '');
```
as the system config input field is a text field, we may need more validation for invalid inputs. like `test|a12|12a|12.4` etc.

MelvinAchterhuis replied:
Yeah I was thinking about that too. Perhaps to prevent invalid values we should also have some sort of validation in the admin when saving?

mitelg replied:
you can also write those values via API, so admin is not enough

MelvinAchterhuis replied:
You're right, I will have a look at the validation in PHP

mitelg commented on src/Storefront/Resources/views/storefront/component/listing/limit.html.twig:
```diff
@@ -0,0 +1,15 @@
+{# @sw-package framework #}
```
what about some blocks?

MelvinAchterhuis replied:
This was inspired by https://github.com/shopware/shopware/blob/trunk/src/Storefront/Resources/views/storefront/component/sorting.html.twig which also doesn't have blocks 😄

mitelg replied:
I mean, fewer blocks means less maintenance for us :man_shrugging: :grin:

MelvinAchterhuis committed:
feat: filter out invalid values, make Twig include logic equal

MelvinAchterhuis committed:
Merge branch 'trunk' into feat/add-products-per-page-dropdown-listing

MelvinAchterhuis commented:
TODO:

- [x] Fix current tests
- [x] Add jest unit test
- [x] Add PHPUnit test
- [x] Add limit handling check in `PagingListingProcessor.php`

MelvinAchterhuis committed:
fix: admin + storefront linting

MelvinAchterhuis committed:
feat: verify request limit against allowed limits

MelvinAchterhuis commented:
I've added the allowedLimits array that is build in `ProductListingCmsElementResolver.php` to the criteria as an extension so that I can compare the requestLimit in `PagingListingProcessor.php` to check if it is valid. Another solution without creating extra queries / dal I couldn't find.

MelvinAchterhuis committed:
feat: add jest test

MelvinAchterhuis committed:
fix: fix method call on null when accessing getVars()

MelvinAchterhuis committed:
fix: fix php-checks

MelvinAchterhuis committed:
feat: add php unit test

feat: add php unit test

mitelg commented on src/Core/Content/Product/Cms/ProductListingCmsElementResolver.php:
```diff
@@ -70,6 +73,12 @@ public function enrich(CmsSlotEntity $slot, ResolverContext $resolverContext, El
+        $this->addListingLimits($slot, $context);
+        $allowedLimits = $slot->getConfig()['limits']['value'] ?? null;
+        if ($allowedLimits !== null) {
+            $criteria->addExtension('limits', new ArrayEntity(['limits' => $allowedLimits]));
```
why not adding a property to the criteria directly?

mitelg commented on src/Core/Content/Product/Cms/ProductListingCmsElementResolver.php:
```diff
@@ -168,4 +177,27 @@ private function restrictFilters(CmsSlotEntity $slot, Request $request): void
+        });
+
+        sort($limits, \SORT_NUMERIC);
+        $slot->setConfig(['limits' => ['value' => $limits]]);
```
this will overwrite existing config

mitelg commented on src/Core/Content/Product/SalesChannel/Listing/Processor/PagingListingProcessor.php:
```diff
@@ -48,19 +48,32 @@ public function process(Request $request, ProductListingResult $result, SalesCha
+    private function getLimit(Criteria $criteria, SalesChannelContext $context, Request $request): int
     {
+        $defaultLimit = $this->config->getInt('core.listing.productsPerPage', $context->getSalesChannelId());
+        $requestLimit = $request->query->has('limit') ? $request->query->getInt('limit') : null;
```
what about limit as request parameter? 

recently we removed setting the limit via request: https://github.com/shopware/shopware/pull/5831

we need to make sure, that only allowed limits are considered. therefore it would be nice, if you could test all cases that "codecov" commented on.

vienthuong replied:
On top of that, if you apply the limit here, doesn't it apply globally for every other slots?

I'd validate and apply the limit in the ProductListingCmsElementResolver so the limit is only applied for specific slot

Maybe something like this in ProductListingCmsElementResolver::enrich

```
        $slots = $resolverContext->getRequest()->query->get('slots') ?? [];
        if (\is_string($slots)) {
            $slots = explode('|', $slots);
        }
        
        if (in_array($slot->getId(), $slots) && $request->query->has('limit')) {
            $this->validateLimit($request, $slot);
            $criteria->setLimit($limit);
        }
```

vienthuong commented on src/Administration/Resources/app/administration/src/module/sw-cms/elements/product-listing/config/sw-cms-el-config-product-listing.html.twig:
```diff
@@ -29,6 +29,14 @@
+                name="pagination"
+                :active-tab="active"
+            >
+                {{ $tc('sw-cms.elements.productListing.config.tab.pagination') }}
```
pagination might be misleading? it could be something like `Listing`

vienthuong commented on src/Administration/Resources/app/administration/src/module/sw-cms/snippet/en-GB.json:
```diff
@@ -358,6 +359,10 @@
             "gridEmptyStateHint": "Try another search term"
+          },
+          "pagination": {
+            "labelLimits": "Limits",
```
Maybe 

```suggestion
            "labelLimits": "Available per page options",
```

vienthuong commented on src/Administration/Resources/app/administration/src/module/sw-cms/snippet/en-GB.json:
```diff
@@ -358,6 +359,10 @@
+          },
+          "pagination": {
+            "labelLimits": "Limits",
+            "helpTextLimits": "The limits are separated by a pipe character."
```
Not sure about this but comma maybe more user friendly than pipe?

vienthuong commented on src/Core/System/Resources/config/listing.xml:
```diff
@@ -72,5 +72,13 @@
+
+        <input-field type="text">
+            <name>limits</name>
+            <label>Limits in listing</label>
```
Available per page listing options

vienthuong commented on src/Core/System/Resources/config/listing.xml:
```diff
@@ -72,5 +72,13 @@
+            <name>limits</name>
+            <label>Limits in listing</label>
+            <label lang="de-DE">Limits in listing</label>
+            <helpText>Limiter for products per page in listing, divided by pipe, for example `6|12|18|24|48`</helpText>
```
Don't we need a migration to set the default config?

MelvinAchterhuis replied:
If we create a migration with a default value the feature is opt-out instead of opt-in, I don't know if that is desirable

vienthuong replied:
Ahh get it 👍🏽, so with no system config & slot config, the dropdown will be hidden? I'd hint it in the help text

marcelbrode commented:
I'm not 100% sure how good this works with viewports. You use numbers, which are multiples of 6, to probably make sure that we have a properly filled listing, when we still have items, but this doesn't work with larger viewports. By default, a normal monitor has 4 Elements per row, so that the option "6" already renders one row and a half, which looks quite weird.

I'd suggest to have multiples of 12 as default (since it would work with 1,2,3 and 4 elements per row). Or do I miss something?

MelvinAchterhuis commented:
> I'm not 100% sure how good this works with viewports. You use numbers, which are multiples of 6, to probably make sure that we have a properly filled listing, if we stillt have items, but this doesn't work with larger viewports. By default, a normal monitor has 4 Elements per row, so that the option "6" already renders one row and a half, which looks quite weird.
> 
> I'd suggest to have multiples of 12 as default (since it would work with 1,2,3 and 4 elements per row). Or do I miss something?

Well I guess that's open for discussion and doesn't impact the implementation that much. We could make the system / element config a multi select instead of a text field that contains the values / multipliers of 12. Advanced users then would still be able to choose their custom values through database or API?

But a side note: I can also set the default products per page to every value I want, that also doesn't always look good on every viewport

marcelbrode commented:
> > I'm not 100% sure how good this works with viewports. You use numbers, which are multiples of 6, to probably make sure that we have a properly filled listing, if we stillt have items, but this doesn't work with larger viewports. By default, a normal monitor has 4 Elements per row, so that the option "6" already renders one row and a half, which looks quite weird.
> > I'd suggest to have multiples of 12 as default (since it would work with 1,2,3 and 4 elements per row). Or do I miss something?
> 
> Well I guess that's open for discussion and doesn't impact the implementation that much. We could make the system / element config a multi select instead of a text field that contains the values / multipliers of 12. Advanced users then would still be able to choose their custom values through database or API?
> 
> But a side note: I can also set the default products per page to every value I want, that also doesn't always look good on every viewport

Since there's lots of other designs, I guess configuring it via text is still a good idea, but the default case, so it looks nice out of the box, should have multiples of 12.

But I guess you're right, so for this new one, we should make it better :D Thanks for work 👍

MelvinAchterhuis commented:
> > > I'm not 100% sure how good this works with viewports. You use numbers, which are multiples of 6, to probably make sure that we have a properly filled listing, if we stillt have items, but this doesn't work with larger viewports. By default, a normal monitor has 4 Elements per row, so that the option "6" already renders one row and a half, which looks quite weird.
> > > I'd suggest to have multiples of 12 as default (since it would work with 1,2,3 and 4 elements per row). Or do I miss something?
> > 
> > 
> > Well I guess that's open for discussion and doesn't impact the implementation that much. We could make the system / element config a multi select instead of a text field that contains the values / multipliers of 12. Advanced users then would still be able to choose their custom values through database or API?
> > But a side note: I can also set the default products per page to every value I want, that also doesn't always look good on every viewport
> 
> Since there's lots of other designs, I guess configuring it via text is still a good idea, but the default case, so it looks nice out of the box, should have multiples of 12.
> 
> But I guess you're right, so for this new one, we should make it better :D Thanks for work 👍

Clear 👍
If we provide default values in the system config I think we should adjust the logic for when it's visible, currently it is

1. System / element config is filled in
2. Merge default products per page value
3. More then 1 unique value
4. Show dropdown

If we provide default values we essentially skip step 1 and 2 and always show the dropdown when this feature is released, so I would vote for:

1. New system / element config to enable the dropdown (by default (or current installations) disabled?)
2. Default values (12 / 24 / 36 / 48 ?)
3. Merge default products per page value
4. More then 1 unique value
5. Show dropdown

That way we don't break existing themes / storefronts but are still able to provide the default values

MartinKrzykawski commented on src/Core/Content/Product/Cms/ProductListingCmsElementResolver.php:
```diff
@@ -168,4 +177,27 @@ private function restrictFilters(CmsSlotEntity $slot, Request $request): void
         }
     }
+
+    private function addListingLimits(CmsSlotEntity $slot, SalesChannelContext $context): void
```
IMO, the readability of this function could be improved by first retrieving the elementLimit, and falling back to the systemLimit if it’s not set. That way, you only need to check a single $limit in the early return. For example:

```
$limit = $slot->getConfig()['limits']['value'] ?? $this->systemConfigService->get('core.listing.limits', $salesChannelId);

if (!$limit) {
    return;
}

// Your filter and merge logic
```

Also, consider adding the final listing limit directly to the criteria here. That way, there’s no need to call `$slot->getConfig()['limits']['value'] ?? null` again outside this function.

MelvinAchterhuis commented:
I will have a look at the feedback after SCD 😄

marcelbrode commented:
Since we noticed no changes in the last 4 weeks, I'll close this ticket. Feel free to ping me, if you've rebased your branch and provided the changes asked and I will gladly reopen this ticket. Thank you :smile:
